### PR TITLE
TST: add test case of ngroup with NaN value

### LINKED ISSUE #50100: BUG: GroupBy.ngroup dropna=False inconsistency when using Categorical
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this bug exists on the main branch of pandas.

### Reproducible Example

```python
# This returns 0 on any Pandas version. That's as expected.
pd.DataFrame.from_dict({'a': [np.nan], 'b': [1]}).groupby(['a', 'b'], dropna=False).ngroup().iloc[0]

# These return -1 on Pandas 1.4.4, and nan in Pandas 1.5.0 or 1.5.2
# From the docs I'm not sure what is meant to happen. Arguably returning -1 is the right behaviour (since in particular it avoids casting the group numbers to float), so this is a bug.
pd.DataFrame.from_dict({'a': [np.nan], 'b': [1]}).groupby(['a', 'b']).ngroup().iloc[0]
pd.DataFrame.from_dict({'a': pd.Categorical([np.nan]), 'b': [1]}).groupby(['a', 'b']).ngroup().iloc[0]

# This returns -1 on Pandas 1.4.4, 0 on Pandas 1.5.0 and nan on 1.5.2
# I'm pretty sure that the correct answer is 0 (for consistency with the float case above)
pd.DataFrame.from_dict({'a': pd.Categorical([np.nan]), 'b': [1]}).groupby(['a', 'b'], dropna=False).ngroup().iloc[0]
```

### Issue Description

Probably related to the recent categorical/groupby changes e.g. #48702. I'm raising this as a separate issue because I don't see any ticket specifically discussing the impact on `ngroup`.

### Expected Behavior

with `dropna=False` groups containing missing values should get their own `ngroup` number. The `ngroup` should never be a float i.e. the group number may contain -1 if dropna=True and any group contains a missing value.

### Installed Versions

<details>
INSTALLED VERSIONS
------------------
commit           : 8dab54d6573f7186ff0c3b6364d5e4dd635ff3e7
python           : 3.10.8.final.0
python-bits      : 64
OS               : Linux
OS-release       : 4.18.0-348.20.1.el8_5.x86_64
Version          : #1 SMP Thu Mar 10 20:59:28 UTC 2022
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_GB.UTF-8
LOCALE           : en_GB.UTF-8

pandas           : 1.5.2
numpy            : 1.23.4
pytz             : 2022.1
dateutil         : 2.8.2
setuptools       : 65.5.0
pip              : 22.2.2
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : None
pandas_datareader: None
bs4              : None
bottleneck       : None
brotli           : None
fastparquet      : None
fsspec           : None
gcsfs            : None
matplotlib       : None
numba            : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pyreadstat       : None
pyxlsb           : None
s3fs             : None
scipy            : None
snappy           : None
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
zstandard        : None
tzdata           : None

/home/mboling/opt/conda/envs/pandas1.5.2/lib/python3.10/site-packages/_distutils_hack/__init__.py:33: UserWarning: Setuptools is replacing distutils.
  warnings.warn("Setuptools is replacing distutils.")
</details>

### PR DESCRIPTION
**Pull request by josemayer**
- [x] closes #50100
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests).
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added new tests to `GroupBy.ngroup` method, which ensure consistency returning a 0 value.

### CODE REVIEW

File: pandas/tests/groupby/test_groupby.py [hunk 1]
```diff
@@ -3189,6 +3189,15 @@ def test_depr_get_group_len_1_list_likes(test_series, kwarg, value, name, warn):
     tm.assert_equal(result, expected)
 
 
+def test_groupby_ngroup_with_nan():
+    # GH#50100
+    df = DataFrame({"a": Categorical([np.nan]), "b": [1]})
+    g = df.groupby(["a", "b"], dropna=False, observed=False).ngroup()
+    result = g.iloc[0]
+    expected = 0
+    tm.assert_equal(result, expected)
```
Comment (mroeschke) [inline pandas/tests/groupby/test_groupby.py:3198]:
```suggestion
    assert result == expected
```

Comment (rhshadrach) [reply pandas/tests/groupby/test_groupby.py]: Can we go the opposite way here: test the full result of ngroup.

```
result = df.groupby(["a", "b"], dropna=False, observed=False).ngroup()
expected = ...
tm.assert_series_equal(result, expected)
```

Comment (mroeschke) [reply pandas/tests/groupby/test_groupby.py]: Agreed, Could you convert the test to this @josemayer

Comment (josemayer) [reply pandas/tests/groupby/test_groupby.py]: Sure!

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @josemayer

**Pull request merged**